net/mptcp/mptcp_sched.c

```diff
@@ -15,58 +22,79 @@ static struct defsched_priv *defsched_get_priv(const struct tcp_sock *tp)
 	return (struct defsched_priv *)&tp->mptcp->mptcp_sched[0];
 }
 
-/* If the sub-socket sk available to send the skb? */
-static bool mptcp_is_available(struct sock *sk, const struct sk_buff *skb,
-			       bool zero_wnd_test)
+static bool mptcp_is_def_unavailable(struct sock *sk)
+{
+	const struct tcp_sock *tp = tcp_sk(sk);
+
+	/* Set of TCP states for which we are not allowed to send data
+	 * "definetively"
+	 */
+	if ((1 << sk->sk_state) & (TCPF_FIN_WAIT1 | TCPF_FIN_WAIT2 |
+				   TCPF_CLOSE | TCPF_CLOSING | TCPF_LISTEN))
+		return true;
+
+	/* The socket failed */
+	if (tp->pf)
+		return true;
+
+	return false;
+}
+
+static bool mptcp_is_temp_unavailable(struct sock *sk,
+				      const struct sk_buff *skb,
+				      bool zero_wnd_test)
 {
 	const struct tcp_sock *tp = tcp_sk(sk);
 	unsigned int mss_now, space, in_flight;
 
-	/* Set of states for which we are allowed to send data */
-	if (!mptcp_sk_can_send(sk))
-		return false;
+	/* Set of TCP states for which we are not allowed to send data
+	 * "temporally"
+	 */
+	if (sk->sk_state == TCPF_SYN_SENT && !tcp_passive_fastopen(sk))
+		return true;
+
+	if ((1 << sk->sk_state) &
+	    (TCPF_SYN_RECV | TCPF_TIME_WAIT | TCPF_LAST_ACK))
```

Comment:
SYN_RECV state is reached when a SYN is received by the host and once the three way-handsake is finished with the reception of the ACK, the TCP connection passes to ESTAB state. Perhaps, TCP-sockets are not externally exposed but I understand that this state must be managed by TCP-stack to reach the ESTAB state.

In any case, in mptcp_ctrl.c and mptcp_output.c there are also checks of the TCPF_SYN_RECV state.

Perhaps, the safest procedure is to check also this state just in case. If so, I would check it in the temporal unavaible-set instead of the definetive unavaible-set.